### ISSUE TITLE
HttpListener Cookie issue

### ISSUE THREAD

**Issue by sawachika-kenji** [dotnet/corefx#22910]
This is my code :

var cookies=new string[]{
"SP_GUEST=; domain=xx.com; path=/; expires=Thu, 01-Jan-1970 00:00:01 GMT",
"SP_F=sadfsdfsdf; domain=xx.com; path=/; expires=Sat, 02-Sep-2017 08:41:07 GMT",
"SP_RETURN_USER=ssdfsdf; domain=xx.com; path=/; expires=Sat, 02-Sep-2017 08:41:07 GMT"
}
 foreach (var c in cookies)
{
  HttpListenerResponse.Headers.Add("Set-Cookie",c);
}

In windows the response cookie is :

Set-Cookie: SP_GUEST=; domain=xx.com; path=/; expires=Thu, 01-Jan-1970 00:00:01 GMT
Set-Cookie: SP_F=sadfsdfsdf; domain=xx.com; path=/; expires=Sat, 02-Sep-2017 08:41:07 GMT
Set-Cookie: SP_RETURN_USER=ssdfsdf; domain=xx.com; path=/; expires=Sat, 02-Sep-2017 08:41:07 GMT

but in Centos7.3 return:

Set-Cookie: SP_GUEST=; domain=xx.com; path=/; expires=Thu, 01-Jan-1970 00:00:01 GMT, SP_F=sadfsdfsdf; domain=xx.com; path=/; expires=Sat, 02-Sep-2017 08:41:07 GMT, SP_RETURN_USER=ssdfsdf; domain=xx.com; path=/; expires=Sat, 02-Sep-2017 08:41:07 GMT

cookie connected together

dotnet --version
2.0.0-preview2-006497

**Comment by Clockwork-Muse** [comment]
Given [this rfc](https://tools.ietf.org/html/rfc6265#section-3) (not sure if superseded):
 >   Origin servers SHOULD NOT fold multiple Set-Cookie header fields into a single header field.  The usual mechanism for folding HTTP headers fields (i.e., as defined in [RFC2616]) might change the semantics of the Set-Cookie header field because the %x2C (",") character is used by Set-Cookie in a way that conflicts with such folding.

... would seem to indicate the Centos version is in the wrong (probably missing something special casing to not fold the header), although I'm not able to find what that header does use the comma for....

**Comment by sawachika-kenji** [comment]
problem still in release 2.0

**Comment by davidsh** [comment]
cc: @stephentoub

**Comment by rmkerr** [comment]
I'm able to repro the behavior that you're seeing on Unix. Interestingly though, the Unix behavior appears to be the behavior that we expect in the cookie tests. While I know for sure that we do not fully support RFC 6265 (which forbids combining set-cookie headers), I'm still interested in tracking down any potential differences between the Windows and Linux implementations. I'll update here if I am able to find anything.

**Comment by rmkerr** [comment]
I did a bit more testing and got a very interesting result. I'd like to hear from @davidsh on what the correct behavior here might be.

In the relevant cookie tests, we add cookies to the `HttpListenerResponse` like this:
```
response.Cookies = cookies;
```
On Unix and on Windows that yields the following Set-Cookie header:
```
Set-Cookie: name1=value1, name2=value2
```
However, in this case the user is instead setting the headers by adding them directly:
```
foreach (var c in cookies)
{
    HttpListenerResponse.Headers.Add("Set-Cookie",c);
}
```
According to the [relevant documentation](https://docs.microsoft.com/en-us/dotnet/api/system.net.webheadercollection.add?view=netcore-2.0#System_Net_WebHeaderCollection_Add_System_String_System_String_) for `WebHeaderCollection.Add` headers added more than once will be added in a comma separated list. So, even adding the headers directly, we expect the same result. That is true on Unix, and we get the following header:
```
Set-Cookie: name1=value1,name2=value2
```
_However_, on Windows we see an unexpected result:
```
Set-Cookie: name1=value1
Set-Cookie: name2=value2
```
While both behaviors are acceptable, I'm not sure which is better. The combined cookies headers seem to be more in line with the behavior we describe in our documentation, but they are discouraged in the latest cookie spec. I think it would be good to decide on a consistent behavior, and to add additional tests to cover both methods of adding cookies.

**Comment by davidsh** [comment]
>While both behaviors are acceptable, I'm not sure which is better. 

Both forms of 'Set-Cookie' headers (1 header w/ 2 cookies vs. 2 headers w/ 1 cookie each) should be interpreted by a client as semantically equivalent.  Neither form is better than than other form 

However, client need to be able to handle either from.   RFC 6265 added this restriction (one cookie per 'Set-Cookie' header) presumably because many user-agents (clients) might parse multiple cookies on a single 'Set-Cookie' header incorrectly. This especially might happen when the cookies have multiple attributes such as an Expires attribute with dates (that have commas in them).

**Comment by rmkerr** [comment]
That sounds reasonable. Since both are valid, are you okay with the inconsistency between Windows and Unix here?

**Comment by davidsh** [comment]
>are you okay with the inconsistency between Windows and Unix here?

Who is this question directed to?  Me or the original person that opened the bug?

Does this difference in behavior actually break clients?

Why is there a difference between the Linux and Windows versions of HttpListener?  Don't both use the same implementation of WebHeaderCollection?

**Comment by rmkerr** [comment]
Sorry for not being clear on that! I meant to ask you David. I assume that the behavior difference was reported here because it breaks clients. I took a look at the code that differs between platforms, and found this comment that describes why Windows splits up the headers:
https://github.com/dotnet/corefx/blob/3db29b721a45fb4f300e27c5df5750f57b38ec45/src/System.Net.HttpListener/src/System/Net/Windows/HttpListenerResponse.Windows.cs#L460-L482 
Just based off of that I think that the Windows behavior is the one we should go with. It seems to give users more control, and is definitely intentional. If you disagree let me know, but otherwise I will be looking into how to ensure that same behavior occurs on Unix. This code also definitely needs some tests!

**Comment by davidsh** [comment]
>I assume that the behavior difference was reported here because it breaks clients.

There is nothing reported here in this issue that says it breaks anything.  The issue simply reported that there was a difference.

If we want to get consistency, then yes, following the single cookie per 'Set-Cookie' header that the Windows (http.sys) implementation is doing should be the goa.

**Comment by rmkerr** [comment]
I think that the fix is simple enough that it's worth pushing for consistency here. I've put something together and will send out a PR soon. If it's not a fix you'd like to make we can discuss there.

**Comment by sawachika-kenji** [comment]
> Set-Cookie: name1=value1,name2=value2

In Google Chrome the cookie name is "name1" and value is "value1,name2=value2", this is obviously wrong, 
the correct should be:
cookie1: name is "name1" and value is "value1"
cookie2: name is "name2" and value is "value2"

**Comment by sawachika-kenji** [comment]
[root@ ~]# dotnet --version
2.0.0
centos 7

This has nothing to do with systems and programs, this is for rfc6265:

> Both forms of 'Set-Cookie' headers (1 header w/ 2 cookies vs. 2 headers w/ 1 cookie each) should be interpreted by a client as semantically equivalent. Neither form is better than than other form.

> Origin servers SHOULD NOT fold multiple Set-Cookie header fields into
   a single header field.  The usual mechanism for folding HTTP headers
   fields (i.e., as defined in [RFC2616]) might change the semantics of
   the Set-Cookie header field because the %x2C (",") character is used
   by Set-Cookie in a way that conflicts with such folding.

**Comment by caesar1995** [comment]
Could you try using .NET Core 2.1?

Does the Set-Cookie header contains `Expires` attribute? If so, I think the issue is fixed here: #26143 for 2.1
